euphoria-spark/src/main/java/cz/seznam/euphoria/spark/SparkFlowTranslator.java

```diff
@@ -51,114 +52,101 @@
-    Translation.set(translations, FlatMap.class, new FlatMapTranslator());
-    Translation.set(translations, ReduceStateByKey.class, new ReduceStateByKeyTranslator(flowSettings));
-    Translation.set(translations, Union.class, new UnionTranslator());
+    // ~ basic operators
+    Translation.add(translations, FlowUnfolder.InputOperator.class, new InputTranslator());
+    Translation.add(translations, FlatMap.class, new FlatMapTranslator());
+    Translation.add(translations, ReduceStateByKey.class, new ReduceStateByKeyTranslator(settings));
+    Translation.add(translations, Union.class, new UnionTranslator());
+
+    // ~ derived operators
+    Translation.add(translations, ReduceByKey.class, new ReduceByKeyTranslator(),
+        ReduceByKeyTranslator::wantTranslate);
 
-    // derived operators
-    Translation.set(translations, ReduceByKey.class, new ReduceByKeyTranslator(),
-            ReduceByKeyTranslator::wantTranslate);
+    // ~ batch broadcast join for a very small left side
+    Translation.add(translations, Join.class, new BroadcastJoinTranslator(), o ->
+        o.getHints().contains(JoinHints.broadcastHashJoin())
+            && (o.getWindowing() == null || o.getWindowing() instanceof GlobalWindowing));
   }
 
   @SuppressWarnings("unchecked")
-  public List<DataSink<?>> translateInto(Flow flow) {
-    // transform flow to acyclic graph of supported operators
-    DAG<Operator<?, ?>> dag = flowToDag(flow);
+  List<DataSink<?>> translateInto(Flow flow) {
+    // ~ transform flow to direct acyclic graph of supported operators
+    final DAG<Operator<?, ?>> dag = flowToDag(flow);
 
-    SparkExecutorContext executorContext =
-            new SparkExecutorContext(sparkEnv, dag, accumulatorFactory, settings);
+    final SparkExecutorContext executorContext =
+        new SparkExecutorContext(sparkEnv, dag, accumulatorFactory, settings);
 
-    // translate each operator to proper Spark transformation
+    // ~ translate each operator to proper Spark transformation
     dag.traverse().map(Node::get).forEach(op -> {
-      Translation tx = translations.get(op.getClass());
-      if (tx == null) {
+      final List<Translation> txs = translations.get(op.getClass());
+      if (txs.isEmpty()) {
         throw new UnsupportedOperationException(
-                "Operator " + op.getClass().getSimpleName() + " not supported");
+            "Operator " + op.getClass().getSimpleName() + " not supported");
       }
       // ~ verify the flowToDag translation
-      Preconditions.checkState(
-              tx.accept == null || Boolean.TRUE.equals(tx.accept.apply(op)));
-
-      JavaRDD<?> out = tx.translator.translate(op, executorContext);
-
-      // save output of current operator to context
-      executorContext.setOutput(op, out);
+      Translation firstMatch = null;
+      for (Translation tx : txs) {
+        if (tx.accept == null || Boolean.TRUE.equals(tx.accept.apply(op))) {
```

Comment:
Why not just `|| tx.accept.apply(op)`? Although I'm aware that this code wasn't written in this PR.